pkg/core/tekton/tekton.go

```diff
@@ -0,0 +1,405 @@
+package tekton
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/pkg/apis"
+
+	"github.com/fuseml/fuseml-core/gen/codeset"
+	"github.com/fuseml/fuseml-core/gen/workflow"
+	"github.com/fuseml/fuseml-core/pkg/core/tekton/builder"
+)
+
+// WorkflowBackend implements the FuseML WorkflowBackend interface for tekton
+type WorkflowBackend struct {
+	namespace     string
+	tektonClients *clients
+}
+
+// NewWorkflowBackend initializes Tekton backend
+func NewWorkflowBackend(namespace string) *WorkflowBackend {
+	return &WorkflowBackend{namespace, newClients(namespace)}
+}
+
+// CreatePipeline receives a FuseML workflow and creates a Tekton pipeline from it
+func (t *WorkflowBackend) CreatePipeline(ctx context.Context, logger *log.Logger, workflow *workflow.Workflow) error {
+	pipeline := generatePipeline(*workflow, t.namespace)
+	logger.Printf("Creating tekton pipeline for workflow: %s...", workflow.Name)
+	_, err := t.tektonClients.PipelineClient.Create(ctx, pipeline, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RunPipeline creates a PipelineRun with its default values for received workflow and codeset
+func (t *WorkflowBackend) RunPipeline(ctx context.Context, workflowName string, codeset codeset.Codeset) error {
+	pipeline, err := t.tektonClients.PipelineClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	pipelineRun, err := generatePipelineRun(pipeline, codeset)
+	if err != nil {
+		return err
+	}
+
+	_, err = t.tektonClients.PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateListener creates tekton resources required to have a listener ready for triggering the pipeline
+func (t *WorkflowBackend) CreateListener(ctx context.Context, logger *log.Logger, workflowName string) (string, error) {
+
+	pipeline, err := t.tektonClients.PipelineClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	triggerTemplate := generateTriggerTemplate(pipeline)
+	_, err = t.tektonClients.TriggerTemplateClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
```

Comment:
Yes, I thought about the idea of doing an update, but then a lot of different situations came to my mind on how it can get complicated and break things that I decided to simple not cover any case of update here.

If you think that we should go further on it here I can starting thinking about it.

Reply:
We don't have to go further on it here, just to summarize what we need to cover in the future. These are immediately obvious, let me know if you have more:
* cleanup in case of partial failure during "create". This should be an "all-or-nothing" type of deal: you either configure everything (listeners, triggers, pipeline etc.) needed for the workflow, or you return an error and nothing is configured in k8s in relation with the workflow.
* implement delete operation (currently missing)
* implement update operation (also missing)